### ISSUE TITLE
Keras reports a warning in the "generative_adversarial_network" notebook and it fails to train.

### ISSUE THREAD

**Issue by tromgy** [the-deep-learners/deep-learning-illustrated#2]
When running "generative_adversarial_network" notebook on the "apple" dataset (downloaded from Google cloud and renamed **apple.npy** to match the path in the notebook), I see this warning:

```
/opt/conda/lib/python3.6/site-packages/keras/engine/training.py:478: UserWarning: Discrepancy between trainable weights and collected trainable weights, did you set `model.trainable` without calling `model.compile` after ?
  'Discrepancy between trainable weights and collected trainable'
```

and the network fails to train as these two images show:

<img width="1129" alt="gan-epoch-99" src="https://user-images.githubusercontent.com/12632548/60056200-160df480-96ae-11e9-9dda-52415dd6ba03.png">

<img width="1119" alt="gan-epoch-1999" src="https://user-images.githubusercontent.com/12632548/60056213-2625d400-96ae-11e9-8daa-58f77c044343.png">

So, trying to get rid of the warning I added the following code:

```
discriminator.compile(loss='binary_crossentropy', 
                      optimizer=RMSprop(lr=0.0008, 
                                        decay=6e-8, 
                                        clipvalue=1.0), 
                      metrics=['accuracy'])
```
right _after_ 
```
discriminator.trainable = False
```
(which was in cell 18) but that just made the network learn making _very sharp_ noise:

<img width="673" alt="recompiled-epoch-99" src="https://user-images.githubusercontent.com/12632548/60056468-0511b300-96af-11e9-8b16-5bc92ecb0a1d.png">

<img width="668" alt="recompiled-gan-epoch-1999" src="https://user-images.githubusercontent.com/12632548/60056485-10fd7500-96af-11e9-8dca-712c1fb564ee.png">

So it seemed that generator was learning, but discriminator was not, which was not surprising because we had never turned the `trainable` parameter back on again!

I then tried to introduce the toggle inside the loop in the `train` function, so that it would set `discriminator.trainable = true`, recompile discriminator, train it, and then set `discriminator.trainable = false` and again recompile discriminator. But that process leads to unbounded memory growth, and the kernel crashes around the 150th iteration (I gave Docker 14 GB), so I was never able to see if there was any change between 99th and 199th steps.

**Comment by tromgy** [comment]
Ok, so the idea with the "toggle" was apparently off the mark. I misunderstood the architecture and how it operates.

Trainability should only be turned off once and before compiling, but I think it needs to apply to the instance of discriminator within the "stacked" adversarial network, and not to the "main" discriminator:

<img width="680" alt="change" src="https://user-images.githubusercontent.com/12632548/60096628-6e310f00-971f-11e9-9064-fd782e7b3689.png">

This avoids the warning, and the network trains fine, but only sometimes! I found that run after run do not produce consistent results, even with exactly same hyper parameters. Obviously there's a random element in the "noise", but here's what I found:

- if the images at the 100th epoch are mostly _black_, the network will train and produce some apple-like results

- but if the images at the 100th epoch are mostly _gray_ (looking like asteroid surface), the network will not train and produce similar images throughout all epochs.

so basically if the network hasn't learned anything in the first hundred iterations, it will never learn anything afterwards.

**Comment by jonkrohn** [comment]
Ah yeah. This is an annoying one. The Keras warning is trying to be helpful because in most cases you wouldn't want to use model weights that haven't been compiled. In a GAN, however, that's _precisely_ what we'd like to do! Check out [here](https://github.com/keras-team/keras/issues/8585) or [here](https://github.com/tensorflow/tensorflow/issues/22012) for more detailed explanations. In older versions of Keras, this warning didn't exist, hence why you don't see any warnings output in my demo [GAN notebook](https://github.com/the-deep-learners/deep-learning-illustrated/blob/master/notebooks/generative_adversarial_network.ipynb).

I'm surprised to hear that the model can train properly if you comment out `discriminator.trainable = False` and set `pred.trainable = False` instead: `pred` isn't a model object, it's a tensor of predicted values. Thus, setting `pred.trainable = False` should only add a new attribute (`trainable`) to the `pred` tensor and have no impact on the discriminator model's weights. 

Now, it's time for me to surprise _you_: @grantbey and I have, over the years, determined empirically that training GANs on CPUs only or on CPUs with a TPU rarely (if ever) leads to the GAN producing the desired output (apples, in this case), instead producing noise (that you aptly described as an asteroid surface). Training GANs on CPUs with a _G_PU, however, reliably produces the desired output (e.g., apples) the majority of the time (in a minority of trials even GPUs produce asteroid surface-y noise). Someday, I'd like to write a detailed blog post about this bizarre phenomenon. It doesn't make any sense because matrix algebra and other computations should be mathematically identical regardless of the device they're computed on.

I'm going to close this issue, but I'm also going to ASAP add a couple of notes to the GAN Jupyter notebook: 

1. That the warning can be safely ignored. 
2. Instructions for running this code in the Google Colab environment, which provides free access to high-performance GPUs.

**Comment by jonkrohn** [comment]
Hey @tromgy -- if you'd like to send me an email at jon@untapt.com with your name and shipping address, I'd like to see if I can get Pearson to ship you a physical copy of _Deep Learning Illustrated_ as thanks for the tremendous help you've been on ironing out these code issues.

**Comment by jonkrohn** [comment]
Actually... if you're able to do that today @tromgy , there might still be time for me to add your name to the _Acknowledgements_ section at the start of the book 😃

**Comment by tromgy** [comment]
Sure, I'll be flattered :)

Alex Lipatov
72 Eddy Rd
Eagle Bridge, NY 12057.

**Comment by tromgy** [comment]
Thanks for the detailed explanation @jonkrohn 

What I found reliably worked on my CPU-only setup are _extremely small_ dataset and correspondingly small batch size:

<img width="623" alt="small-set-small-batch" src="https://user-images.githubusercontent.com/12632548/60139359-90587a80-977b-11e9-8017-675e89700c37.png">

makes it train quickly, too! I also tested on the "cars" dataset and it worked there as well.

**Comment by jonkrohn** [comment]
Oh wow that is super interesting... what happens with the full dataset but still a very small batch size like `5`? The potential issue with the small data set is that there isn't a rich set of examples that the generator can use to learn how to produce detailed, diverse images with.

**Comment by tromgy** [comment]
Using full dataset even with a small batch still produces gray images all the way through. So I think it might be the opposite -- the more variation in the set, the harder it is for the network to learn. At least with these doodles this seems to be the case.

**Comment by jonkrohn** [comment]
Yeah that makes sense... except I still don't understand why the results of any computation would be different on a CPU relative to a GPU... From everything that I know, the results should be identical on a CPU-only machine except that it should take longer to produce those results 😕

**Comment by jonkrohn** [comment]
Ok, my friend @tromgy -- I've pushed an updated GAN notebook. If you get a chance, please check out the notes I added.
